nodeadm/internal/util/sysinfo.go

```diff
@@ -0,0 +1,286 @@
+		}
+		if !isCPUOnline(cpuDir) {
+			continue
+		}
+
+		rawPhysicalID, err := getCoreID(cpuDir)
+		if os.IsNotExist(err) {
+			zap.L().Warn(fmt.Sprintf("Cannot read core id for %s, core_id file does not exist, err: %s", cpuDir, err))
+			continue
+		} else if err != nil {
+			return nil, err
+		}
+		physicalID, err := strconv.Atoi(rawPhysicalID)
+		if err != nil {
+			return nil, err
+		}
+
+		rawPhysicalPackageID, err := getCPUPhysicalPackageID(cpuDir)
+		if os.IsNotExist(err) {
+			zap.L().Warn(fmt.Sprintf("Cannot read physical package id for %s, physical_package_id file does not exist, err: %s", cpuDir, err))
+			continue
+		} else if err != nil {
+			return nil, err
+		}
+
+		physicalPackageID, err := strconv.Atoi(rawPhysicalPackageID)
+		if err != nil {
+			return nil, err
+		}
+
+		coreIDx := -1
+		for id, core := range cores {
+			if core.Id == physicalID && core.SocketID == physicalPackageID {
+				coreIDx = id
+			}
+		}
+		if coreIDx == -1 {
+			cores = append(cores, Core{})
+			coreIDx = len(cores) - 1
+		}
+		desiredCore := &cores[coreIDx]
+
+		desiredCore.Id = physicalID
+		desiredCore.SocketID = physicalPackageID
+
+		if len(desiredCore.Threads) == 0 {
+			desiredCore.Threads = []int{cpuID}
+		} else {
+			desiredCore.Threads = append(desiredCore.Threads, cpuID)
+		}
+
+	}
+	return cores, nil
+}
+
+func getMatchedInt(rgx *regexp.Regexp, str string) (int, error) {
+	matches := rgx.FindStringSubmatch(str)
+	if len(matches) != 2 {
+		return 0, fmt.Errorf("failed to match regexp, str: %s", str)
+	}
+	valInt, err := strconv.Atoi(matches[1])
+	if err != nil {
+		return 0, err
+	}
+	return valInt, nil
+}
+
+func getCoreID(cpuPath string) (string, error) {
+	coreIDFilePath := fmt.Sprintf("%s%s", cpuPath, coreIDFilePath)
+	// #nosec G304 // This cpuPath essentially come from getCPUsPaths and it should be a system path
+	coreID, err := os.ReadFile(coreIDFilePath)
+	if err != nil {
+		return "", err
+	}
+	return strings.TrimSpace(string(coreID)), err
+}
+
+func isCPUOnline(cpuPath string) bool {
+	cpuOnlinePath, err := filepath.Abs(cpusPath + "online")
+	if err != nil {
+		zap.L().Info(fmt.Sprintf("Unable to get absolute path for %s", cpuPath))
+		return false
+	}
+
+	// Quick check to determine if file exists: if it does not then kernel CPU hotplug is disabled and all CPUs are online.
+	_, err = os.Stat(cpuOnlinePath)
+	if err != nil && os.IsNotExist(err) {
+		return true
+	}
+	if err != nil {
+		zap.L().Info(fmt.Sprintf("Unable to stat %s: %s", cpuOnlinePath, err))
+	}
+
+	cpuID, err := getCPUID(cpuPath)
+	if err != nil {
+		zap.L().Info(fmt.Sprintf("Unable to get CPU ID from path %s: %s", cpuPath, err))
+		return false
+	}
+
+	isOnline, err := isCpuIDOnline(cpuOnlinePath, cpuID)
+	if err != nil {
+		zap.L().Info(fmt.Sprintf("Unable to get online CPUs list: %s", err))
+		return false
+	}
+	return isOnline
+}
+
+func getCPUID(dir string) (uint16, error) {
+	regex := regexp.MustCompile("cpu([0-9]+)")
```

Comment:
is this effectively the same as `cpuDirRegExp` on line 14?

Reply:
good catch, I took a closer look, and we can remove `getCPUID` entirely and use `getMatchedInt`